### ISSUE TITLE
Mongodb system is down and unable to recover when the featureCompatibilityVersion is not specified and changed to an invalid value

### ISSUE THREAD

**Issue by hoyhbx** [mongodb/mongodb-kubernetes-operator#1072]
**What did you do to encounter the bug?**
I changed the featureCompatibilityVersion to an invalid value `4.2.4`. This caused all mongodb-agent containers to remain in an unready state. When a valid CR is applied in an attempt to recover state, the operator is unable to create the replica set due to the previously invalid value and the CR deployment eventually fails and has to be manually redeployed.

It is very important to resolve this issue as it greatly hurts the availability of the cluster because the entire cluster has to be deleted and then redeployed which causes a very long system downtime. This cannot be afforded as high availability is central to MongoDB’s guarantees.

1. Apply the following CR spec without specifying a featureCompatibilityVersion:

```yaml
spec:
  members: 3
  security:
    authentication:
      modes:
      - SCRAM
  statefulSet:
    spec:
      template:
        spec:
          containers:
          - name: mongod
            resources:
              limits:
                cpu: '0.2'
                memory: 250M
              requests:
                cpu: '0.2'
                memory: 200M
          - name: mongodb-agent
            resources:
              limits:
                cpu: '0.2'
                memory: 250M
              requests:
                cpu: '0.2'
                memory: 200M
```

2. Add the `featureCompatibilityVersion` to the spec with an invalid value:

```yaml
spec:
  featureCompatibilityVersion: ‘4.2.4’
```
3. Reapply the original CR in Step 1 to attempt to revert, but the operator is stuck.

**What did you expect?**
The operator should raise an error for the invalid input and/or replace it with an empty string when the invalid CR is applied. The state should be recovered when a valid CR is applied such that all containers in the ReplicaSet are ready and the operator reconciles successfully.

**What happened instead?**
The mongodb-agent containers in all pods remain in an unready state. The operator keeps requeuing reconciliation requests. When a valid CR is applied, the operator fails to build the automation config with the following error:

```bash
2022-08-08T05:06:56.052Z    ERROR    controllers/mongodb_status_options.go:104    Error deploying MongoDB ReplicaSet: failed to ensure AutomationConfig: could not build automation config: could not create an automation config: can't build the automation config: can't compute semver version from previous FeatureCompatibilityVersion 4.2.4
```

The CR deployment eventually fails:

```bash
NAME           PHASE    VERSION
test-cluster   Failed   4.4.0

```

**Screenshots**

**Operator Information**
- Operator Version - `0.7.3`
- MongoDB Image used - `4.4.0`

**Kubernetes Cluster Information**
<details>
<summary><code>kubectl version --short --output=yaml</code></summary>
<br><pre>
$ kubectl version --short --output=yaml
clientVersion:
  buildDate: "2022-05-24T12:26:19Z"
  compiler: gc
  gitCommit: 3ddd0f45aa91e2f30c70734b175631bec5b5825a
  gitTreeState: clean
  gitVersion: v1.24.1
  goVersion: go1.18.2
  major: "1"
  minor: "24"
  platform: linux/amd64
kustomizeVersion: v4.5.4
serverVersion:
  buildDate: "2021-05-21T23:01:33Z"
  compiler: gc
  gitCommit: 5e58841cce77d4bc13713ad2b91fa0d961e69192
  gitTreeState: clean
  gitVersion: v1.21.1
  goVersion: go1.16.4
  major: "1"
  minor: "21"
  platform: linux/amd64
</pre>
</details>

**Additional context**
- **mongodb-agent container**: var/log/mongodb-mms-automation/readiness.log
```bash
test-cluster-0:/$ tail var/log/mongodb-mms-automation/readiness.log
{"level":"info","ts":1659633518.9905996,"msg":"Agent has reached goal state"}
{"level":"info","ts":1659633529.0694976,"msg":"Agent has reached goal state"}
{"level":"info","ts":1659633539.0994744,"msg":"Agent has reached goal state"}
{"level":"info","ts":1659633549.079082,"msg":"Agent has reached goal state"}
{"level":"info","ts":1659633559.0674996,"msg":"Agent has reached goal state"}
{"level":"info","ts":1659633568.998716,"msg":"Agent has reached goal state"}
{"level":"info","ts":1659633579.082997,"msg":"Agent has reached goal state"}
{"level":"debug","ts":1659633588.992177,"msg":"The Agent hasn't reported working on the new config yet, the last plan finished at 2022-08-04T16:51:41Z"}
{"level":"debug","ts":1659633599.0842354,"msg":"The Agent hasn't reported working on the new config yet, the last plan finished at 2022-08-04T16:51:41Z"}
{"level":"debug","ts":1659633609.0684364,"msg":"The Agent hasn't reported working on the new config yet, the last plan finished at 2022-08-04T16:51:41Z"}
```

- **mongodb-agent container**: var/log/mongodb-mms-automation/healthstatus/agent-health-status.json
```json
{"statuses":{"test-cluster-0":{"IsInGoalState":false,"LastMongoUpTime":1659633835,"ExpectedToBeUp":true,"ReplicationStatus":2}},"mmsStatus":{"test-cluster-0":{"name":"test-cluster-0","lastGoalVersionAchieved":1,"plans":[{"started":"2022-08-04T16:50:55.426418702Z","completed":"2022-08-04T16:51:07.815756371Z","moves":[{"move":"Start","moveDoc":"Start the process","steps":[{"step":"StartFresh","stepDoc":"Start a mongo instance  (start fresh)","isWaitStep":false,"started":"2022-08-04T16:50:55.42646342Z","completed":"2022-08-04T16:51:06.429702683Z","result":"success"}]},{"move":"WaitAllRsMembersUp","moveDoc":"Wait until all members of this process' repl set are up","steps":[{"step":"WaitAllRsMembersUp","stepDoc":"Wait until all members of this process' repl set are up","isWaitStep":true,"started":"2022-08-04T16:51:06.429775278Z","completed":null,"result":"wait"}]},{"move":"RsInit","moveDoc":"Initialize a replica set including the current MongoDB process","steps":[{"step":"RsInit","stepDoc":"Initialize a replica set","isWaitStep":false,"started":null,"completed":null,"result":""}]},{"move":"WaitFeatureCompatibilityVersionCorrect","moveDoc":"Wait for featureCompatibilityVersion to be right","steps":[{"step":"WaitFeatureCompatibilityVersionCorrect","stepDoc":"Wait for featureCompatibilityVersion to be right","isWaitStep":true,"started":null,"completed":null,"result":""}]}]},{"started":"2022-08-04T16:51:39.819161369Z","completed":"2022-08-04T16:51:41.418194096Z","moves":[{"move":"WaitHasCorrectAutomationCredentials","moveDoc":"Wait for the automation user to be added (if needed)","steps":[{"step":"WaitHasCorrectAutomationCredentials","stepDoc":"Wait for the automation user to be added (if needed)","isWaitStep":true,"started":"2022-08-04T16:51:39.81918117Z","completed":null,"result":"wait"}]}]},{"started":"2022-08-04T17:20:45.382688107Z","completed":null,"moves":[{"move":"WaitFeatureCompatibilityVersionCorrect","moveDoc":"Wait for featureCompatibilityVersion to be right","steps":[{"step":"WaitFeatureCompatibilityVersionCorrect","stepDoc":"Wait for featureCompatibilityVersion to be right","isWaitStep":true,"started":"2022-08-04T17:20:45.382706615Z","completed":null,"result":"wait"}]}]}],"errorCode":0,"errorString":""}}}
```

- **mongodb-agent container log**
```bash
[2022-08-04T16:26:51.487+0000] [.info] [src/director/director.go:planAndExecute:563] <test-cluster-0> [16:26:51.487] Step=WaitCanUpdateFeatureCompatibilityVersion as part of Move=WaitCanUpdateFeatureCompatibilityVersion in plan failed : <test-cluster-0> [16:26:51.487] Postcondition not yet met for step WaitCanUpdateFeatureCompatibilityVersion because
['currentState.Fickle.CanUpdateFeatureCompatibilityVersion' = false].
 Recomputing a plan...
[2022-08-04T16:26:52.120+0000] [.info] [src/config/config.go:ReadClusterConfig:437] [16:26:52.120] Retrieving cluster config from /var/lib/automation/config/cluster-config.json...
[2022-08-04T16:26:52.120+0000] [.info] [main/components/agent.go:LoadClusterConfig:270] [16:26:52.120] clusterConfig unchanged
[2022-08-04T16:26:52.514+0000] [.info] [src/mongoclientservice/mongoclientservice.go:func1:1573] [16:26:52.514] Testing auth with username __system db=local to test-cluster-0.test-cluster-svc.mongodb.svc.cluster.local:27017 (local=false) connectMode=SingleConnect ipversion=0 tls=false
[2022-08-04T16:26:52.686+0000] [.info] [src/mongoctl/processctl.go:GetKeyHashes:2026] <test-cluster-0> [16:26:52.686] Able to successfully auth to test-cluster-0.test-cluster-svc.mongodb.svc.cluster.local:27017 (local=false) using desired auth key
[2022-08-04T16:26:52.981+0000] [.info] [src/mongoctl/processctl.go:Update:3307] <test-cluster-0> [16:26:52.981] <DB_WRITE> Updated with query map[] and update [{$set [{agentFeatures [StateCache]} {nextVersion 2}]}] and upsert=true on local.clustermanager
[2022-08-04T16:26:53.174+0000] [.info] [src/config/config.go:ReadClusterConfig:437] [16:26:53.174] Retrieving cluster config from /var/lib/automation/config/cluster-config.json...
[2022-08-04T16:26:53.175+0000] [.info] [main/components/agent.go:LoadClusterConfig:270] [16:26:53.175] clusterConfig unchanged
[2022-08-04T16:26:53.181+0000] [.info] [src/director/director.go:computePlan:279] <test-cluster-0> [16:26:53.181] ... process has a plan : WaitCanUpdateFeatureCompatibilityVersion,UpdateFeatureCompatibilityVersion
[2022-08-04T16:26:53.181+0000] [.info] [src/director/director.go:tracef:793] <test-cluster-0> [16:26:53.181] Running step: 'WaitCanUpdateFeatureCompatibilityVersion' of move 'WaitCanUpdateFeatureCompatibilityVersion'
[2022-08-04T16:26:53.181+0000] [.info] [src/director/director.go:tracef:793] <test-cluster-0> [16:26:53.181] because
[All the following are true:
    ['currentState.Up' = true]
    ['currentState.Version = 4.4.0-563487e100c4215e2dce98d0af2a6a5a2d67c5cf() is equal to desiredState.Version = 4.4.0-()' = true]
]
[2022-08-04T16:26:53.182+0000] [.info] [src/director/director.go:planAndExecute:563] <test-cluster-0> [16:26:53.182] Step=WaitCanUpdateFeatureCompatibilityVersion as part of Move=WaitCanUpdateFeatureCompatibilityVersion in plan failed : <test-cluster-0> [16:26:53.182] Postcondition not yet met for step WaitCanUpdateFeatureCompatibilityVersion because
['currentState.Fickle.CanUpdateFeatureCompatibilityVersion' = false].
 Recomputing a plan...
```
- **operator log**
```bash
2022-08-04T15:42:37.184Z    INFO    controllers/replica_set_controller.go:140    Reconciling MongoDB    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:37.184Z    DEBUG    controllers/replica_set_controller.go:142    Validating MongoDB.Spec    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:37.185Z    DEBUG    controllers/replica_set_controller.go:151    Ensuring the service exists    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:37.204Z    INFO    controllers/replica_set_controller.go:475    Create/Update operation succeeded    {"ReplicaSet": "mongodb/test-cluster", "operation": "updated"}
2022-08-04T15:42:37.204Z    DEBUG    controllers/replica_set_controller.go:160    Ensuring the service for Arbiters exists    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:37.210Z    INFO    controllers/replica_set_controller.go:475    Create/Update operation succeeded    {"ReplicaSet": "mongodb/test-cluster", "operation": "updated"}
2022-08-04T15:42:37.211Z    INFO    controllers/replica_set_controller.go:371    Creating/Updating AutomationConfig    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:37.236Z    DEBUG    scram/scram.go:166    Credentials have not changed, using credentials stored in: secret/my-scram-scram-credentials
2022-08-04T15:42:37.236Z    DEBUG    controllers/replica_set_controller.go:525    AutomationConfigMembersThisReconciliation    {"mdb.AutomationConfigMembersThisReconciliation()": 3}
2022-08-04T15:42:37.236Z    DEBUG    controllers/replica_set_controller.go:394    Waiting for agents to reach version 2    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:37.236Z    DEBUG    agent/agent_readiness.go:65    The Agent in the Pod 'test-cluster-0' hasn't reached the goal state yet (goal: 2, agent: 1)    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:37.236Z    INFO    controllers/mongodb_status_options.go:110    ReplicaSet is not yet ready, retrying in 10 seconds
2022-08-04T15:42:41.846Z    INFO    controllers/replica_set_controller.go:140    Reconciling MongoDB    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:41.846Z    DEBUG    controllers/replica_set_controller.go:142    Validating MongoDB.Spec    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:41.846Z    DEBUG    controllers/replica_set_controller.go:151    Ensuring the service exists    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:41.854Z    INFO    controllers/replica_set_controller.go:475    Create/Update operation succeeded    {"ReplicaSet": "mongodb/test-cluster", "operation": "updated"}
2022-08-04T15:42:41.854Z    DEBUG    controllers/replica_set_controller.go:160    Ensuring the service for Arbiters exists    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:41.859Z    INFO    controllers/replica_set_controller.go:475    Create/Update operation succeeded    {"ReplicaSet": "mongodb/test-cluster", "operation": "updated"}
2022-08-04T15:42:41.859Z    INFO    controllers/replica_set_controller.go:371    Creating/Updating AutomationConfig    {"ReplicaSet": "mongodb/test-cluster"}
2022-08-04T15:42:41.890Z    DEBUG    scram/scram.go:166    Credentials have not changed, using credentials stored in: secret/my-scram-scram-credentials
2022-08-04T15:42:41.891Z    DEBUG    controllers/replica_set_controller.go:525    AutomationConfigMembersThisReconciliation    {"mdb.AutomationConfigMembersThisReconciliation()": 3}
2022-08-04T15:42:41.893Z    ERROR    controllers/mongodb_status_options.go:104    Error deploying MongoDB ReplicaSet: failed to ensure AutomationConfig: could not build automation config: could not create an automation config: can't build the automation config: can't compute semver version from previous FeatureCompatibilityVersion vcvxvmalbd
github.com/mongodb/mongodb-kubernetes-operator/controllers.messageOption.ApplyOption
    /workspace/controllers/mongodb_status_options.go:104
github.com/mongodb/mongodb-kubernetes-operator/pkg/util/status.Update
    /workspace/pkg/util/status/status.go:25
github.com/mongodb/mongodb-kubernetes-operator/controllers.ReplicaSetReconciler.Reconcile
    /workspace/controllers/replica_set_controller.go:212
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
    /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
    /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
    /root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.3/pkg/internal/controller/controller.go:227
```

### Possible root cause and fix

While building the AutomationConfig, as the previous `featureCompatibilityVersion` is invalid, the semver version is unable to be computed as part of [`b.setFeatureCompatibilityVersionIfUpgradeIsHappening()`](https://github.com/mongodb/mongodb-kubernetes-operator/blob/1acfbdf5465e589386e11cae8455b83dac20c37e/pkg/automationconfig/automation_config_builder.go#L236).

A simple fix would be to validate the `featureCompatibilityVersion` input. This would prevent the reduced availability and subsequent failure of the deployed replica set demonstrated above.

Validation can be performed by using a SemVer Parser similar to the one used in `CalculateFeatureCompatibilityVersion`.

**Comment by hoyhbx** [comment]
Remove stale

**Comment by dan-mckean** [comment]
Hi, I'm Dan and I'm the Product Manager for MongoDB's support of Kubernetes.

I'm doing some work right now to try and identify how the Community Operator is being used. The Community Operator is something I inherited when I started at MongoDB, but it doesn't get as much attention from us as we'd like and we're trying to understand how it's used in order to establish its future. It will help us prioritize future issues and PRs raised by the community 🙂

Here's a super short survey (it's much easier for us to review all the feedback that way!): [https://docs.google.com/forms/d/e/1FAIpQLSfwrwyxBSlUyJ6AmC-eYlgW_3JEdfA48SB2i5--_WpiynMW2w/viewform?usp=sf_link](https://docs.google.com/forms/d/e/1FAIpQLSfwrwyxBSlUyJ6AmC-eYlgW_3JEdfA48SB2i5--_WpiynMW2w/viewform?usp=sf_link)

 If you'd rather email me instead: [dan.mckean@mongodb.com](mailto:dan.mckean@mongodb.com?subject=MongoDB%20Community%20Operator%20feedback)

Thank you in advance!
Dan
